[Accordion] Add Joy UI Accordion components

siriwatknp opened the pull request:
Preview: https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-accordion/

closes #36281. We have to roll a custom implementation as Base UI isn't available #38037 yet.

The Accordion family reuses styles from the List family.

```
AccordionGroup <-> List
Accordion <-> ListItem
AccordionSummary <-> ListItemButton
```

Joy UI Accordion uses CSS Grid for the accordion detail to let developers use CSS transition to animate the open/closed state.

---

oliviertassinari commented on docs/data/joy/components/accordion/AccordionTransition.js:
```diff
@@ -0,0 +1,86 @@
+      >
+        <Radio value="0.3s linear" label="Linear" />
+        <Radio value="0.4s ease" label="Easing" />
+        <Radio value="0.8s cubic-bezier(0.68, -0.6, 0.32, 1.6)" label="Bouncy" />
```
It takes 277ms for the accordion to start moving, I don't think that we can't have this option, per https://web.dev/fid/#what-is-a-good-fid-score we are close to "Poor". For comparison, with Easing, I measure this at 27ms.

```suggestion
```

Instead, I think that we could have a small bounce at the end of the animation like https://dribbble.com/shots/20711653-Side-panel-framework-for-Opkey-s-no-code-automation-tool?

oliviertassinari commented on docs/data/joy/components/accordion/AccordionTransition.js:
```diff
@@ -0,0 +1,86 @@
+        <Radio value="mix" label="Mix" />
+      </RadioGroup>
+      <AccordionGroup
+        transition={
```
Would it be clearer if this was customized with CSS and the sx prop? Developer would have direct access to their CSS animation design tokens, like duration, easing, etc.

What problem does the `transition` prop solve better?

siriwatknp replied:
I thought about it as well. With `transition` prop, developers don't have to remember and specify both properties `grid-template-rows` and `padding`.

Also, if you want to customize the transition for both initial and expanded state in CSS, you will need to do this:

```js
<AccordionGroup
  sx={{
    '--AccordionDetails-transition': `grid-template-rows 0.2s, padding-block 0.2s`,
        [`& .${accordionDetailsClasses.root}.${accordionDetailsClasses.expanded}`]: {
          '--AccordionDetails-transition': `grid-template-rows 0.3s, padding-block 0.3s`,
        },
  }}
/>
```

whereas with transition prop:
```js
<AccordionGroup
  transition={{
    initial: '0.2s',
    expanded: '0.3s',
  }}
/>
```

oliviertassinari commented on docs/data/joy/components/accordion/AccordionSizes.js:
```diff
@@ -0,0 +1,40 @@
+  const [size, setSize] = React.useState('md');
+  return (
+    <Stack spacing={2} sx={{ maxWidth: 400, flex: 1 }}>
+      <ToggleButtonGroup
+        size="sm"
+        buttonFlex={1}
+        value={size}
+        onChange={(event, newValue) => setSize(newValue)}
```
There can be no value selected, expected?

<img width="446" alt="Screenshot 2023-07-27 at 15 35 58" src="https://github.com/mui/material-ui/assets/3165635/165fed37-8cfe-4bb9-bb9a-05478d6252aa">

siriwatknp replied:
How to reproduce it?

oliviertassinari replied:
https://github.com/mui/material-ui/assets/3165635/99e6bc94-1197-457d-8c4f-acf53928807d

https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-accordion/#sizes

Oh, actually https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-toggle-button-group/#editor-toolbar has the same bug now? One of the text alignment must be active at all time.

siriwatknp replied:
Hmm, I don't think so. We should let developers handle by themselve.

oliviertassinari commented on docs/data/joy/components/accordion/AccordionUsage.js:
```diff
@@ -0,0 +1,68 @@
+import AccordionDetails from '@mui/joy/AccordionDetails';
+import AccordionSummary from '@mui/joy/AccordionSummary';
+
+export default function ButtonUsage() {
```
I think that we will need to do something about this demo 😁 https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-accordion/#introduction

https://github.com/mui/material-ui/assets/3165635/99360748-eac4-4257-9bb1-2a975b993b29

oliviertassinari replied:
Changing the alignment for the demo in the playground to be aligned at the top will likely solve this.

oliviertassinari reviewed (commented):
Great to see movement on the accordion 👍 

I wonder about the order we create this vs. the one in Base UI.

More ideas to push it further:

1. For the animation duration, I think that `getAutoHeightDuration` could make a difference in how great it feels, Material UI is following this duration curve based on the height, to control the transition px/s
   
<img width="502" alt="Screenshot 2023-07-27 at 15 31 52" src="https://github.com/mui/material-ui/assets/3165635/6b23e95d-467c-4331-9da0-1ef61d7d8ada">

2. We will likely need `overflowAnchor`
3. I think that we will need to consider the performance implication of mounting the whole accordion details list.
4. This makes me think that we could start to support https://github.com/mui/material-ui/issues/16367

adding @DiegoAndai to the loop, as the [owner of the Accordion component](https://www.notion.so/mui-org/Components-Surfaces-a478c993928149768044f689a9864edc).

oliviertassinari commented on docs/data/joy/components/accordion/AccordionTransition.js:
```diff
@@ -0,0 +1,86 @@
+          setTransition(event.target.value);
+        }}
+      >
+        <Radio value="0.3s linear" label="Linear" />
```
This easing makes me feel uncomfortable. What if instead, we were slowing the default animation? Or have it to be instant?

siriwatknp replied:
I think the default animation/transition is out of scope from this PR that's why the demo illustrates how to add transitions. 

@zanivan mentioned the default transition for Joy as well and I think it should be done separately so that each component does not need to wait to roll out.

zanivan replied:
I think adding default animations/transitions to Joy UI would be really helpful. It could improve the microinteractions on components and potentially become a standout feature that sets Joy UI apart from competitors. Also, if we incorporate the Joy UI design language into Base UI, it would further differentiate the two products.

All in all, I think we can polish out a bit these animations, the easing is indeed a bit off, mostly on the bouncing demo—at least for me. Are there any workarounds to improve these animations before coming up with the default transition?

oliviertassinari replied:
> I think the default animation/transition is out of scope from this PR that's why the demo illustrates how to add transitions.

@siriwatknp Ah sorry, I used the wrong terminology. By "default animation", I meant the one applied on the demo.

But actually, this connects well to point 1. in https://github.com/mui/material-ui/pull/38164#pullrequestreview-1549831929. This feels OK, maybe a 10% faster would be nice

https://github.com/mui/material-ui/assets/3165635/1c145b3a-66d5-4970-837f-c3bfb0e38335

This feels too fast; 40% slower would be great:

https://github.com/mui/material-ui/assets/3165635/4073ac39-6e8e-4899-9d54-3d78f9f56100

oliviertassinari replied:
I think we can also see another bug in this animation, the children isn't stable, "The Accordion Group supports the…" moves, so to start reading the content, you have to wait for the whole transition to finish

https://github.com/mui/material-ui/assets/3165635/e6f91ae3-2287-43ad-b742-22e1c505eb77

It's better in https://ui.shadcn.com/docs/components/accordion.

siriwatknp commented:
> I wonder about the order we create this vs. the one in Base UI.

For the Accordion to match WAI ARIA guide, using an internal hook from `@mui/utils` is enough.

> We will likely need overflowAnchor

Got it.

> I think that we will need to consider the performance implication of mounting the whole accordion details list.

Why? I think there will be a rare case to think in terms of performance which I think we can leave to the developer to decide.

oliviertassinari commented:
> Why? I think there will be a rare case to think in terms of performance which I think we can leave to the developer to decide.

@siriwatknp I think it's the same problem as in https://github.com/mui/material-ui/issues/21250. Covered in https://mui.com/material-ui/react-accordion/#performance.

siriwatknp commented:
> I think it's the same problem as in #21250. I think that the default of mounting everything is best, agree. The point I lead to is: How can developers unmount what's not visible with Joy UI without breaking animations?

I see. That's a valid point. Based on the priority, I would that it is an enhancement that we can find later to not block the release of the new component.

zanivan commented:
Hey @siriwatknp, I think we could polish out a bit the accordion component, and I have a few takes on it:
- The outlined variant would make more sense if it had the dividers by default
- When expanded, we could have a bit of extra padding between the `AccordionDetails` and the `AccordionSummary` below. When more items are expanded, is a bit hard to group them visually
![Screenshot 2023-08-09 at 15 50 31](https://github.com/mui/material-ui/assets/37222944/1bc4dcba-033f-4f73-bbce-1ed90181e782)
Or, we can leave the divider by default on all variants
![Screenshot 2023-08-09 at 15 52 12](https://github.com/mui/material-ui/assets/37222944/4b09ded1-368b-4619-a110-ac5005de3c89)

- I would remove the `solid` variant, or go with `invertedColors` by default—at least on this variant
- The size variations `md` and `lg` could have the same font sizes, just variating the paddings
- I wouldn't change the horizontal padding on different sizes, this could disrupt the users' grids and at the same time makes the large variant to have less space than the smaller ones.
- I really like the arrow size on the `sm` variant, and I'd keep it on the `md` as well
- I believe a default animation would be great, even if it's just a simple CSS animation like we have on the demo `<AccordionGroup transition="0.2s ease"> ` 


All in all, all these are ideas, and I'd love to hear your thoughts on it as well. Anyway, it's a great step forward, and one more component to the lib 🎉

siriwatknp commented:
@zanivan Updates:

- The divider appears by default which can be hidden by `<AccordionGroup disableDivider>` prop
- The default transition is `0.2s ease`.
- Adjusted padding-bottom to increase spacing between AccordionSummary and Details between accordions.
- I am hesitating to remove the `solid` variant at this point. I feel like it's the presenting issue more than the implementation so I decided to remove the Usage demo for now.
- I did not touch size-related properties. I would not add more CSS to it at this point because you can simply customize the font-size with just one line.
- Updated the last common example. The `AccordionDetails` should have only one for each accordion.

zanivan commented:
- I don't know if we should remove the `Usage demo`, since we have it in all interactive components, it would be inconsistent.
So this is a matter of choosing what tradeoff we will have: removing the `Usage demo` that'll result in an inconsistency with other pages; or, keeping the `Usage demo` and the `solid` variant. 
Personally, I lean towards keeping the `Usage demo`, since we have more components with problematic variants like this—and this issue can be addressed in a separate PR, where we select which variants each component will have.

- The https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-accordion/#styling-on-expansion doesn't seem to be rendering the border on the third item
![Screenshot 2023-08-10 at 08 41 59](https://github.com/mui/material-ui/assets/37222944/390863ad-a0a6-40ba-a30c-3a77b7e3962d)

oliviertassinari reviewed (commented):
A quick review:

- We could add the playground back, at least, why not.
- In the size demo, I was a bit surprised to see the typography change as well, I don't know, I feel should we want to control spacing and typography size independently.
- Rebasing on HEAD could be nice, to pull in the latest changes
- https://github.com/mui/material-ui/issues/38037#issuecomment-1678190180
- I linked this PR to the Base UI issue, so that the community, if they decide to take on Base UI accordion can more easily find the work we did here.
- I wish we had more demo previews. Maybe 3 accordion isn't compatible with this (too much vertical space), but it would work with 2.
- We could add `ComponentLinkHeader`
- https://github.com/mui/material-ui/issues/38037#issuecomment-1678200974

DiegoAndai commented on packages/mui-joy/src/AccordionDetails/AccordionDetailsProps.ts:
```diff
@@ -0,0 +1,67 @@
+export interface AccordionDetailsPropsColorOverrides {}
+
+export interface AccordionDetailsTypeMap<P = {}, D extends React.ElementType = 'div'> {
+  props: P & {
```
There's an [ongoing effort](https://github.com/mui/material-ui/pull/38171) to move away from one letter type parameters, I think we should enforce it in Joy for new components. We should also create a PR in the future for existing one letter type parameters.

siriwatknp replied:
Thanks @DiegoAndai

siriwatknp commented on packages/mui-joy/src/AccordionSummary/accordionSummaryClasses.ts:
```diff
@@ -0,0 +1,27 @@
+  button: string;
+  /** Class name applied to the indicator element. */
+  indicator: string;
+  /** Class name applied when the accordion is disabled. */
+  disabled: string;
+  /** Class name applied when the accordion is expanded. */
+  expanded: string;
```
For styling purposes, without these classes, you will have to theme the `AccordionSummary` like this:

```js
extendTheme({
  components: {
    JoyAccordion: {
      styleOverrides: {
        root: {
          [`.${accordionClasses.root}.${accordionClasses.expanded} &`]: {
             …styles
          }
        }
      }
    }
  }
})
```

whereas 
```js
extendTheme({
  components: {
    JoyAccordion: {
      styleOverrides: {
        root: {
          [`&.${accordionSummaryClasses.expanded}`]: {
             …styles
          }
        }
      }
    }
  }
})
```

DiegoAndai replied:
I agree it's easier, but if we do it for the `AccordionSummary` component, shouldn't we do it for the other accordion components as well?

- `AccordionDetails` has `expanded` but not `disabled`
- `AccordionGroup` has neither

We should either add the classes to all or none. Otherwise, it might be confusing.

I would only have these on the root, I think the following is not a bad API:

```js
extendTheme({
  components: {
    JoyAccordion: {
      styleOverrides: {
        root: {
          [`.${accordionClasses.expanded} &.${accordionSummaryClasses.root}`]: {
             …styles
          }
        }
      }
    }
  }
})
```

siriwatknp commented:
> * I don't know if we should remove the `Usage demo`, since we have it in all interactive components, it would be inconsistent.
>   So this is a matter of choosing what tradeoff we will have: removing the `Usage demo` that'll result in an inconsistency with other pages; or, keeping the `Usage demo` and the `solid` variant.
>   Personally, I lean towards keeping the `Usage demo`, since we have more components with problematic variants like this—and this issue can be addressed in a separate PR, where we select which variants each component will have.
> * The https://deploy-preview-38164--material-ui.netlify.app/joy-ui/react-accordion/#styling-on-expansion doesn't seem to be rendering the border on the third item
>   ![Screenshot 2023-08-10 at 08 41 59](https://user-images.githubusercontent.com/37222944/259733518-390863ad-a0a6-40ba-a30c-3a77b7e3962d.png)

I brought back the Usage demo and apply the variant to other components when it is solid. Also, fix the demo you mentioned.

siriwatknp commented on packages/mui-joy/src/Accordion/Accordion.test.tsx:
```diff
@@ -0,0 +1,145 @@
+    expect(container.firstChild).to.have.class(classes.disabled);
+  });
+
+  it('should warn when switching from controlled to uncontrolled', () => {
+    const wrapper = render(
+      <Accordion expanded>
+        <AccordionSummary>Header</AccordionSummary>
+      </Accordion>,
+    );
+
+    expect(() => wrapper.setProps({ expanded: undefined })).to.toErrorDev(
+      'MUI: A component is changing the controlled expanded state of Accordion to be uncontrolled.',
+    );
+  });
+
+  it('should warn when switching between uncontrolled to controlled', () => {
+    const { setProps } = render(
+      <Accordion>
+        <AccordionSummary>Header</AccordionSummary>
+      </Accordion>,
+    );
+
+    expect(() => setProps({ expanded: true })).toErrorDev(
+      'MUI: A component is changing the uncontrolled expanded state of Accordion to be controlled.',
+    );
+  });
```
I'd wait a bit more. There are not many components that implement this warning. Adding too many things to describeConformance could make it too complicated.

zanivan reviewed (approved):
Added some small tweaks to a demo, but it looks good to me! 🚀
